lib/FreeRTOS/freertos-plus-mqtt/freertos_mqtt_agent.c

```diff
@@ -1432,43 +1444,57 @@ MQTTStatus_t MQTTAgent_Init( MQTTContextHandle_t mqttContextHandle,
 
 /*-----------------------------------------------------------*/
 
-MQTTContext_t * MQTTAgent_CommandLoop( void )
+MQTTStatus_t MQTTAgent_CommandLoop( MQTTContextHandle_t * pOutContextHandle,
+                                    MQTTContext_t ** pOutContext )
 {
     Command_t * pCommand;
     MQTTStatus_t operationStatus = MQTTSuccess;
-    MQTTContext_t * ret = NULL;
     CommandType_t currentCommandType = NONE;
 
     /* The command queue should have been created before this task gets created. */
     configASSERT( commandQueue );
 
+    if( ( pOutContextHandle == NULL ) || ( pOutContext == NULL ) )
+    {
+        operationStatus = MQTTBadParameter;
+    }
+
     /* Loop until we receive a terminate command. */
-    for( ; ; )
+    for( ; operationStatus == MQTTSuccess; )
     {
         /* Wait for the next command, if any. */
         pCommand = NULL;
         xQueueReceive( commandQueue, &( pCommand ), pdMS_TO_TICKS( MQTT_AGENT_MAX_EVENT_QUEUE_WAIT_TIME ) );
         /* Set the command type in case the command is released while processing. */
         currentCommandType = ( pCommand ) ? pCommand->commandType : NONE;
-        operationStatus = processCommand( pCommand, &ret );
+        operationStatus = processCommand( pCommand, pOutContext );
 
-        /* Return the current MQTT context if status was not successful. */
-        if( operationStatus != MQTTSuccess )
+        /* Return the current MQTT context on disconnect or error. */
+        if( ( currentCommandType == DISCONNECT ) || ( operationStatus != MQTTSuccess ) )
         {
-            LogError( ( "MQTT operation failed with status %s\n",
-                        MQTT_Status_strerror( operationStatus ) ) );
+            /* Retrieve context handle from context. */
+            configASSERT( ( *pOutContext >= mqttContexts ) &&
+                          ( *pOutContext <= &mqttContexts[ MQTT_AGENT_MAX_SIMULTANEOUS_CONNECTIONS - 1 ] ) );
```

Comment:
*pOutContext is of type `MQTTContext_t *` but `&mqttContexts[ MQTT_AGENT_MAX_SIMULTANEOUS_CONNECTIONS - 1 ]` is a double pointer right? Should the `&` be removed?

Reply:
`mqttContexts` is an array of `MQTTContext_t`. The `[]` gives the last element, and the `&` returns its address, so it's only a single pointer. `(mqttContexts + MQTT_AGENT_MAX_SIMULTANEOUS_CONNECTIONS - 1)` is what we want